plugins/hls-code-range-plugin/src/Ide/Plugin/CodeRange.hs

```diff
@@ -77,49 +77,62 @@ foldingRangeHandler ide _ FoldingRangeParams{..} = do
+        foldingRanges <- liftIO . runAction "FoldingRange" ide $
             getFoldingRanges filePath
         pure . List $ foldingRanges
     where
         uri :: Uri
         TextDocumentIdentifier uri = _textDocument
 
-getFoldingRanges :: NormalizedFilePath -> ExceptT String IdeAction [FoldingRange]
-getFoldingRanges file = do
-    (codeRange, _) <- maybeToExceptT "fail to get code range" $ useE GetCodeRange file
-
-    -- removing first node because it folds the entire file
-    pure $ drop 1 $ findFoldingRanges codeRange
+getFoldingRanges :: NormalizedFilePath -> Action [FoldingRange]
+getFoldingRanges file = fmap (maybe [] findFoldingRanges) $ use GetCodeRange file
 
 selectionRangeHandler :: IdeState -> PluginId -> SelectionRangeParams -> LspM c (Either ResponseError (List SelectionRange))
 selectionRangeHandler ide _ SelectionRangeParams{..} = do
     pluginResponse $ do
         filePath <- ExceptT . pure . maybeToEither "fail to convert uri to file path" $
                 toNormalizedFilePath' <$> uriToFilePath' uri
-        selectionRanges <- ExceptT . liftIO . runIdeAction "SelectionRange" (shakeExtras ide) . runExceptT $
-            getSelectionRanges filePath positions
-        pure . List $ selectionRanges
+        fmap List . mapExceptT runIdeAction' . getSelectionRanges filePath $ positions
   where
     uri :: Uri
     TextDocumentIdentifier uri = _textDocument
 
     positions :: [Position]
     List positions = _positions
 
-getSelectionRanges :: NormalizedFilePath -> [Position] -> ExceptT String IdeAction [SelectionRange]
+    runIdeAction' :: IdeAction (Either SelectionRangeError [SelectionRange]) -> LspT c IO (Either String [SelectionRange])
+    runIdeAction' action = do
+        result <- liftIO $ runIdeAction "SelectionRange" (shakeExtras ide) action
+        pure $ case result of
+            Left err   -> maybe (Right []) Left (showError err)
+            Right list -> Right list
+
+    showError :: SelectionRangeError -> Maybe String
+    -- This might happen if the HieAst is not ready, so we give it a default value instead of throwing an error
+    showError SelectionRangeBadDependency = Nothing
+    showError SelectionRangeInputPositionMappingFailure = Just "failed to apply position mapping to input positions"
+    showError SelectionRangeOutputPositionMappingFailure = Just "failed to apply position mapping to output positions"
+
+data SelectionRangeError = SelectionRangeBadDependency
+                         | SelectionRangeInputPositionMappingFailure
+                         | SelectionRangeOutputPositionMappingFailure
+
+getSelectionRanges :: NormalizedFilePath -> [Position] -> ExceptT SelectionRangeError IdeAction [SelectionRange]
 getSelectionRanges file positions = do
-    (codeRange, positionMapping) <- maybeToExceptT "fail to get code range" $ useE GetCodeRange file
+    (codeRange, positionMapping) <- maybeToExceptT SelectionRangeBadDependency . MaybeT $
+        useWithStaleFast GetCodeRange file
```

Comment:
maybe we should include the dependency that failed in the error so we can log it later? I guess `useWithStaleFast` doesn't log if it fails?

Reply:
Right, I realised when thinking about this that a plugin returning a `ResponseError` can be "normal" and not necessarily something that HLS needs to log as an error. I still think it's good to log them so you know what's happening, but maybe it doesn't need to be an error log.